autotuning/evd/gen_evd_block_vectorized.py

```diff
@@ -0,0 +1,200 @@
+
+def load_intrinsic(n, mat, var, size, inv):
+    load_list = []
+    for i in range(n):
+        j = str(i)
+        load_list.append(mat + "_" + var + j + " = " + "_mm256_load_pd(" + mat + " + " + \
+                         size + " * " + var + " + " + inv + " + " + str(i * 4) + ");")
+    return load_list
+
+def gen_evd_col_operations(n, mat, size, inv):
+    instr_list = []
+    instr_list.extend(vec_declarations(n, [mat]))
+    instr_list.extend(vec_init_col(n, mat, size, inv))
+    instr_list.extend(two_opr_intrinsic(n, "mul", "sin_col", mat+"_col", "sin_vec"))
+    instr_list.extend(three_opr_intrinsic(n, "fmsub", mat+"_row", mat+"_row", "cos_vec", "sin_col"))
+    instr_list.extend(vec_store_col(n, mat, "row", size, inv))
+    instr_list.extend(two_opr_intrinsic(n, "mul", "sin_row", mat+"_rcopy", "sin_vec"))
+    instr_list.extend(three_opr_intrinsic(n, "fmadd", mat+"_col", mat+"_col", "cos_vec", "sin_row"))
+    instr_list.extend(vec_store_col(n, mat, "col", size, inv))
+    return instr_list
+
+def gen_evd_row_operations(n, mat_list, size, inv):
+    instr_list = []
+    instr_list.extend(vec_declarations(n, mat_list))
+    for mat in mat_list:
+        instr_list.extend(vec_init_row(n, mat, size, inv))
+        instr_list.extend(two_opr_intrinsic(n, "mul", "sin_col", mat+"_col", "sin_vec"))
+        instr_list.extend(three_opr_intrinsic(n, "fmsub", mat+"_row", mat+"_row", "cos_vec", "sin_col"))
+        instr_list.extend(vec_store_row(n, mat, "row", size, inv))
+        instr_list.extend(two_opr_intrinsic(n, "mul", "sin_row", mat+"_rcopy", "sin_vec"))
+        instr_list.extend(three_opr_intrinsic(n, "fmadd", mat+"_col", mat+"_col", "cos_vec", "sin_row"))
+        instr_list.extend(vec_store_row(n, mat, "col", size, inv))
+    return instr_list
+
+def gen_evd_block_vector(unroll_count):
+    function_code = \
+'''static size_t evd_block_vector(struct matrix_t Amat, struct matrix_t Vmat, size_t epoch) {
+    assert(Amat.rows == Amat.cols);
+    double* A = Amat.ptr;
+
+    double* V = Vmat.ptr;
+    const size_t m = Amat.rows;
+
+    matrix_identity(Vmat);
+
+    for (size_t ep = 1; ep <= epoch; ep++) {
+        double cos_t, sin_t;
+
+        for (size_t row = 0; row < m; row++) {
+            for (size_t col = row + 1; col < m; col++) {
+                size_t n = m;
+                __m256d sin_vec, cos_vec;
+
+                // Compute cos_t and sin_t for the rotation
+                sym_jacobi_coeffs(A[row * m + row], A[row * m + col], A[col * m + col], &cos_t, &sin_t);
+
+                sin_vec = _mm256_set1_pd(sin_t);
+                cos_vec = _mm256_set1_pd(cos_t);
+
+                if (m % URC != 0) n = m - (m % URC);
+
+                // Compute the eigen values by updating the columns until convergence
+                for (size_t i = 0; i < n; i += URC) {
+                    // UNROLLED_COL_OPS
+                }
+
+                if (m % URC != 0) {
+                    for (size_t i = 0; i < m - n; i++) {
```

Comment:
I think a better approach here would be first a loop on the form `for (size_t i = 0; i + UTC - 1< m; i += URC){ //UNROLLED_COL_OPS}` and then a second one on the form `for (; i < m; ++i){ //Handle the leftovers} ` This would remove the need to check the modulo and keeping track of n. It also feels like the current structure could have issues when `m < UTC`.

Reply:
Yes, I tried that, but it doesn't work as expected! We get the same precision error...But it does reduce the extra modulo operation.

Reply:
I think we can consider only two levels of unrolling for the report, since third level of unrolling doesn't result in good performance anyways even if it converges or not.